File: src/lib/fiskalni.ts
import type { SqlDb } from './sqldb';

/** Parse a fiscal receipt number; returns null for refunds (R-...), empty, or non-numeric. */
export function parseFiskalniBroj(raw: string | null | undefined): number | null {
  if (raw == null) return null;
  const t = String(raw).trim();
  if (!/^\d+$/.test(t)) return null;
  return parseInt(t, 10);
}

/** Najviše koliko praznina vraćamo odjednom. */
export const MAX_PRAZNINA = 200;

/**
 * Vrati brojeve koji nedostaju strogo između najmanjeg i najvećeg fiskalnog broja.
 *
 * Rezultat je ograničen na `maxGaps` — jedan pogrešno ukucan broj (npr. 1234567
 * umjesto 1234) inače generiše milione "praznina" i zamrzne ekran koji ih crta.
 */
export function izracunajPraznine(
  brojevi: number[],
  maxGaps: number = MAX_PRAZNINA,
  ignorisani: Set<number> = new Set()
): number[] {
  const present = new Set(brojevi);
  const sorted = [...present].sort((a, b) => a - b);
  if (sorted.length < 2 || maxGaps <= 0) return [];
  const gaps: number[] = [];
  const max = sorted[sorted.length - 1];
  for (let n = sorted[0] + 1; n < max; n++) {
    if (present.has(n) || ignorisani.has(n)) continue;
    gaps.push(n);
    if (gaps.length >= maxGaps) break;
  }
  return gaps;
}

/**
 * Postavka: posljednji BF broj odštampan prije nego je program preuzeo niz.
 * Služi samo dok u bazi nema nijednog fiskalizovanog računa.
 */
export const ZADNJI_FISKALNI_KEY = 'fiscal.zadnjiBroj';

/** Kada je taj broj upisan — odlučuje je li noviji od posljednjeg računa u bazi. */
export const ZADNJI_FISKALNI_AT_KEY = 'fiscal.zadnjiBrojAt';

/**
 * BF sa posljednjeg fiskalizovanog računa; `null` kad ga nema.
 *
 * Uzima se posljednji po datumu računa, a ne najveći broj u bazi: naknadno
 * popunjena praznina u nizu (stari račun ukucan danas) nosi stari datum, pa ne
 * pomjera niz unazad, a resetovan brojač na uređaju ne ostavlja zaglavljen
 * maksimum iz prošlog perioda. Reklamacije (`R-3`) i ručni upisi bez broja se
 * preskaču, pa se gleda prvi račun odozgo koji ima numerički BF.
 */
export function zadnjiFiskalniRacun(db: SqlDb): { broj: number; createdAt: string } | null {
  const rows = db
    .prepare(`
      SELECT brojFiskalnogRacuna, createdAt FROM orders
      WHERE brojFiskalnogRacuna IS NOT NULL
      ORDER BY createdAt DESC, id DESC
      LIMIT 200
    `)
    .all() as Array<{ brojFiskalnogRacuna: string; createdAt: string }>;
  for (const r of rows) {
    const broj = parseFiskalniBroj(r.brojFiskalnogRacuna);
    if (broj !== null) return { broj, createdAt: r.createdAt ?? '' };
  }
  return null;
}

/** Samo broj sa posljednjeg fiskalizovanog računa. */
export function zadnjiFiskalniBroj(db: SqlDb): number | null {
  return zadnjiFiskalniRacun(db)?.broj ?? null;
}

/** Ručno upisan posljednji broj iz postavki; `null` kad nije upisan. */
export function zadnjiUpisaniFiskalniBroj(db: SqlDb): number | null {
  const row = db.prepare('SELECT value FROM settings WHERE key = ?').get(ZADNJI_FISKALNI_KEY) as
    { value: string } | undefined;
  const broj = row ? parseInt(row.value, 10) : NaN;
  return Number.isInteger(broj) && broj >= 0 ? broj : null;
}

/** Trenutak ručnog upisa, u istom formatu kao `orders.createdAt`. */
function zadnjiUpisAt(db: SqlDb): string | null {
  const row = db.prepare('SELECT value FROM settings WHERE key = ?').get(ZADNJI_FISKALNI_AT_KEY) as
    { value: string } | undefined;
  return row?.value ?? null;
}

/**
 * Upisuje posljednji odštampani BF broj i vrijeme upisa. Nula znači „uređaj još
 * nije štampao". Vrijeme je bitno: ručna ispravka mora pobijediti ono što baza
 * zna, inače operater ne može ispraviti niz koji je odlutao.
 */
export function postaviZadnjiFiskalniBroj(db: SqlDb, broj: number): number {
  if (!Number.isInteger(broj) || broj < 0) {
    throw new Error('Posljednji fiskalni broj mora biti cijeli broj 0 ili veći');
  }
  const upsert = db.prepare(
    'INSERT INTO settings (key, value) VALUES (?, ?) ON CONFLICT(key) DO UPDATE SET value = excluded.value'
  );
  upsert.run(ZADNJI_FISKALNI_KEY, String(broj));
  const { sada } = db.prepare("SELECT datetime('now','localtime') AS sada").get() as { sada: string };
  upsert.run(ZADNJI_FISKALNI_AT_KEY, sada);
  return broj;
}

/**
 * Broj koji će uređaj po svoj prilici dati sljedećem isječku.
 *
 * Potreban je računu po prilogu: naziv zbirne stavke se kuca prije štampe, a
 * mora nositi broj tog istog isječka. `null` znači da se broj ne može
 * pretpostaviti i da ga operater mora unijeti; predviđanje nije garancija, pa
 * se poslije štampe poredi sa stvarnim BF-om.
 */
export function predvidjeniFiskalniBroj(db: SqlDb): number | null {
  const racun = zadnjiFiskalniRacun(db);
  const upisani = zadnjiUpisaniFiskalniBroj(db);
  if (upisani === null) return racun ? racun.broj + 1 : null;
  if (!racun) return upisani + 1;
  // Novija činjenica pobjeđuje: ručna ispravka važi dok kroz program ne prođe
  // sljedeći račun, a onda ga baza opet preuzima.
  const upisAt = zadnjiUpisAt(db) ?? '';
  return (upisAt > racun.createdAt ? upisani : racun.broj) + 1;
}

File: src-tauri/backend/src/fiskalni.rs
//! Fiskalni niz: brojevi računa, praznine, predviđeni sljedeći BF (`lib/fiskalni.ts`).

use std::collections::HashSet;

use regex::Regex;
use serde_json::Value;

use crate::greska::R;
use crate::js;
use crate::p;
use crate::sql::Db;

/// Broj fiskalnog računa; `None` za reklamacije (R-...), prazno ili nenumeričko.
pub fn parse_fiskalni_broj(raw: &Value) -> Option<i64> {
    if raw.is_null() {
        return None;
    }
    let t = js::to_string(raw);
    let t = t.trim();
    thread_local!(static CIFRE: Regex = Regex::new(r"^[0-9]+$").unwrap());
    if !CIFRE.with(|r| r.is_match(t)) {
        return None;
    }
    js::parse_int(t)
}

pub const MAX_PRAZNINA: usize = 200;

/// Brojevi koji nedostaju strogo između najmanjeg i najvećeg, najviše `max_gaps`.
pub fn izracunaj_praznine(brojevi: &[i64], max_gaps: usize, ignorisani: &HashSet<i64>) -> Vec<i64> {
    let present: HashSet<i64> = brojevi.iter().copied().collect();
    let mut sorted: Vec<i64> = present.iter().copied().collect();
    sorted.sort();
    if sorted.len() < 2 || max_gaps == 0 {
        return vec![];
    }
    let mut gaps = Vec::new();
    let max = *sorted.last().unwrap();
    let mut n = sorted[0] + 1;
    while n < max {
        if !present.contains(&n) && !ignorisani.contains(&n) {
            gaps.push(n);
            if gaps.len() >= max_gaps {
                break;
            }
        }
        n += 1;
    }
    gaps
}

pub const ZADNJI_FISKALNI_KEY: &str = "fiscal.zadnjiBroj";
pub const ZADNJI_FISKALNI_AT_KEY: &str = "fiscal.zadnjiBrojAt";

/// BF sa posljednjeg fiskalizovanog računa (po datumu, ne najveći broj).
pub fn zadnji_fiskalni_racun(db: &Db) -> R<Option<(i64, String)>> {
    let rows = db.all(
        "SELECT brojFiskalnogRacuna, createdAt FROM orders
      WHERE brojFiskalnogRacuna IS NOT NULL
      ORDER BY createdAt DESC, id DESC
      LIMIT 200",
        p![],
    )?;
    for r in rows {
        if let Some(broj) = parse_fiskalni_broj(&r["brojFiskalnogRacuna"]) {
            let at = r["createdAt"].as_str().unwrap_or("").to_string();
            return Ok(Some((broj, at)));
        }
    }
    Ok(None)
}

pub fn zadnji_fiskalni_broj(db: &Db) -> R<Option<i64>> {
    Ok(zadnji_fiskalni_racun(db)?.map(|(b, _)| b))
}

/// Ručno upisan posljednji broj iz postavki.
pub fn zadnji_upisani_fiskalni_broj(db: &Db) -> R<Option<i64>> {
    let v = db.val("SELECT value FROM settings WHERE key = ?", p![ZADNJI_FISKALNI_KEY])?;
    if v.is_null() {
        return Ok(None);
    }
    Ok(js::parse_int(&js::to_string(&v)).filter(|b| *b >= 0))
}

fn zadnji_upis_at(db: &Db) -> R<Option<String>> {
    let v = db.val("SELECT value FROM settings WHERE key = ?", p![ZADNJI_FISKALNI_AT_KEY])?;
    Ok(if v.is_null() { None } else { Some(js::to_string(&v)) })
}

pub fn postavi_zadnji_fiskalni_broj(db: &Db, broj: &Value) -> R<i64> {
    if !js::is_integer(broj) || broj.as_f64().unwrap() < 0.0 {
        crate::baci!("Posljednji fiskalni broj mora biti cijeli broj 0 ili veći");
    }
    let broj = broj.as_f64().unwrap() as i64;
    let upsert = "INSERT INTO settings (key, value) VALUES (?, ?) ON CONFLICT(key) DO UPDATE SET value = excluded.value";
    db.run(upsert, p![ZADNJI_FISKALNI_KEY, broj.to_string()])?;
    let sada = db.val("SELECT datetime('now','localtime') AS sada", p![])?;
    db.run(upsert, p![ZADNJI_FISKALNI_AT_KEY, sada])?;
    Ok(broj)
}

/// Broj koji će uređaj po svoj prilici dati sljedećem isječku.
pub fn predvidjeni_fiskalni_broj(db: &Db) -> R<Option<i64>> {
    let racun = zadnji_fiskalni_racun(db)?;
    let upisani = zadnji_upisani_fiskalni_broj(db)?;
    let Some(upisani) = upisani else {
        return Ok(racun.map(|(b, _)| b + 1));
    };
    let Some((broj, created)) = racun else {
        return Ok(Some(upisani + 1));
    };
    let upis_at = zadnji_upis_at(db)?.unwrap_or_default();
    Ok(Some(if upis_at > created { upisani } else { broj } + 1))
}
